File: src/test/java/no/fintlabs/flyt/gateway/application/archive/resource/web/CaseSearchParametersServiceTest.java
package no.fintlabs.flyt.gateway.application.archive.resource.web;

import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.resource.arkiv.kodeverk.SaksmappetypeResource;
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource;
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource;
import no.fint.model.resource.arkiv.noark.ArkivdelResource;
import no.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource;
import no.fintlabs.cache.FintCache;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.CaseSearchParametersDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.KlasseDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.SakDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.SkjermingDto;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class CaseSearchParametersServiceTest {

    @Mock
    private FintCache<String, ArkivdelResource> arkivdelResourceCache;

    @Mock
    private FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache;

    @Mock
    private FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache;

    @Mock
    private FintCache<String, SaksmappetypeResource> saksmappetypeResourceCache;

    @Mock
    private FintCache<String, KlassifikasjonssystemResource> klassifikasjonssystemResourceCache;

    @Mock
    private SakDto sakDto;

    private CaseSearchParametersService caseSearchParametersService;

    private CaseSearchParametersDto.CaseSearchParametersDtoBuilder caseSearchParametersDtoBuilder;

    @BeforeEach
    void setUp() {
        caseSearchParametersService = new CaseSearchParametersService(
                arkivdelResourceCache,
                administrativEnhetResourceCache,
                tilgangsrestriksjonResourceCache,
                saksmappetypeResourceCache,
                klassifikasjonssystemResourceCache
        );
        caseSearchParametersDtoBuilder = CaseSearchParametersDto.builder();
    }

    @Test
    void shouldCreateFilterQueryParamValueForArkivdel() {
        CaseSearchParametersDto caseSearchParametersDto = caseSearchParametersDtoBuilder.arkivdel(true).build();

        Identifikator identifikator = new Identifikator();
        identifikator.setIdentifikatorverdi("arkivdelId");

        ArkivdelResource arkivdelResource = new ArkivdelResource();
        arkivdelResource.setSystemId(identifikator);

        when(sakDto.getArkivdel()).thenReturn(Optional.of("arkivdelKey"));
        when(arkivdelResourceCache.get("arkivdelKey")).thenReturn(arkivdelResource);

        String result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto);

        assertEquals("arkivdel eq 'arkivdelId'", result);
    }

    @Test
    void shouldCreateFilterQueryParamValueForAdministrativEnhetAndTilgangsrestriksjon() {
        CaseSearchParametersDto caseSearchParametersDto =
                caseSearchParametersDtoBuilder
                        .administrativEnhet(true)
                        .tilgangsrestriksjon(true)
                        .build();

        Identifikator administrativEnhetIdentifikator = new Identifikator();
        administrativEnhetIdentifikator.setIdentifikatorverdi("adminEnhetId");

        AdministrativEnhetResource administrativEnhetResource = new AdministrativEnhetResource();
        administrativEnhetResource.setSystemId(administrativEnhetIdentifikator);

        Identifikator tilgangsrestriksjonIdentifikator = new Identifikator();
        tilgangsrestriksjonIdentifikator.setIdentifikatorverdi("tilgangsId");

        TilgangsrestriksjonResource tilgangsrestriksjon = new TilgangsrestriksjonResource();
        tilgangsrestriksjon.setSystemId(tilgangsrestriksjonIdentifikator);

        SkjermingDto skjermingDto = SkjermingDto
                .builder()
                .tilgangsrestriksjon("tilgangsKey")
                .build();

        when(sakDto.getAdministrativEnhet()).thenReturn(Optional.of("adminEnhetKey"));
        when(administrativEnhetResourceCache.get("adminEnhetKey")).thenReturn(administrativEnhetResource);

        when(sakDto.getSkjerming()).thenReturn(Optional.of(skjermingDto));
        when(tilgangsrestriksjonResourceCache.get("tilgangsKey")).thenReturn(tilgangsrestriksjon);

        String result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto);

        assertEquals("administrativenhet eq 'adminEnhetId' and tilgangskode eq 'tilgangsId'", result);
    }

    @Test
    void shouldCreateFilterQueryParamValueForKlasseringKlassifikasjonssystem() {
        CaseSearchParametersDto caseSearchParametersDto = caseSearchParametersDtoBuilder
                .klassering(true)
                .klasseringRekkefolge("1")
                .klasseringKlassifikasjonssystem(true)
                .build();

        Identifikator klassifikasjonssystemIdentifikator = new Identifikator();
        klassifikasjonssystemIdentifikator.setIdentifikatorverdi("klassifikasjonssystemId");

        KlassifikasjonssystemResource klassifikasjonssystemResource = new KlassifikasjonssystemResource();
        klassifikasjonssystemResource.setSystemId(klassifikasjonssystemIdentifikator);

        KlasseDto klasseDto = KlasseDto.builder()
                .rekkefolge(1)
                .klassifikasjonssystem("klassifikasjonssystemKey")
                .build();

        when(sakDto.getKlasse()).thenReturn(Optional.of(List.of(klasseDto)));
        when(klassifikasjonssystemResourceCache.get("klassifikasjonssystemKey")).thenReturn(klassifikasjonssystemResource);

        String result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto);

        assertEquals("klassifikasjon/primar/ordning eq 'klassifikasjonssystemId'", result);
    }

    @Test
    void shouldCreateFilterQueryParamValueForKlasseringKlasseId() {
        CaseSearchParametersDto caseSearchParametersDto = caseSearchParametersDtoBuilder
                .klassering(true)
                .klasseringRekkefolge("1")
                .klasseringKlasseId(true)
                .build();

        KlasseDto klasseDto = KlasseDto.builder()
                .rekkefolge(1)
                .klasseId("klasseId")
                .build();

        when(sakDto.getKlasse()).thenReturn(Optional.of(List.of(klasseDto)));

        String result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto);

        assertEquals("klassifikasjon/primar/verdi eq 'klasseId'", result);
    }

}

File: src/test/kotlin/no/novari/flyt/archive/gateway/resource/web/CaseSearchParametersServiceTest.kt
package no.novari.flyt.archive.gateway.resource.web

import no.novari.cache.FintCache
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator
import no.novari.fint.model.resource.arkiv.kodeverk.SaksstatusResource
import no.novari.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource
import no.novari.fint.model.resource.arkiv.noark.AdministrativEnhetResource
import no.novari.fint.model.resource.arkiv.noark.ArkivdelResource
import no.novari.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource
import no.novari.flyt.archive.gateway.dispatch.model.instance.CaseSearchParametersDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.KlasseDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.SkjermingDto
import no.novari.flyt.archive.gateway.resource.web.exceptions.KlasseOrderOutOfBoundsException
import no.novari.flyt.archive.gateway.resource.web.exceptions.SearchKlasseOrderNotFoundInCaseException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

@ExtendWith(MockitoExtension::class)
class CaseSearchParametersServiceTest {
    @Mock
    private lateinit var arkivdelResourceCache: FintCache<String, ArkivdelResource>

    @Mock
    private lateinit var administrativEnhetResourceCache: FintCache<String, AdministrativEnhetResource>

    @Mock
    private lateinit var tilgangsrestriksjonResourceCache: FintCache<String, TilgangsrestriksjonResource>

    @Mock
    private lateinit var saksstatusResourceCache: FintCache<String, SaksstatusResource>

    @Mock
    private lateinit var klassifikasjonssystemResourceCache: FintCache<String, KlassifikasjonssystemResource>

    private lateinit var caseSearchParametersService: CaseSearchParametersService

    @BeforeEach
    fun setUp() {
        caseSearchParametersService =
            CaseSearchParametersService(
                arkivdelResourceCache,
                administrativEnhetResourceCache,
                tilgangsrestriksjonResourceCache,
                mock(),
                saksstatusResourceCache,
                klassifikasjonssystemResourceCache,
            )
    }

    @Test
    fun `creates a filter query param value for arkivdel`() {
        val arkivdel =
            ArkivdelResource().apply {
                systemId = Identifikator().apply { identifikatorverdi = "arkivdelId" }
            }
        whenever(arkivdelResourceCache.get("arkivdelKey")).thenReturn(arkivdel)

        val sakDto = SakDto.builder().arkivdel("arkivdelKey").build()
        val caseSearchParametersDto = CaseSearchParametersDto.builder().arkivdel(true).build()

        val result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto)

        assertThat(result).isEqualTo("arkivdel eq 'arkivdelId'")
    }

    @Test
    fun `creates a filter query param value for administrativEnhet and tilgangsrestriksjon`() {
        val administrativEnhetResource =
            AdministrativEnhetResource().apply {
                systemId =
                    Identifikator().apply { identifikatorverdi = "adminEnhetId" }
            }
        val tilgangsrestriksjonResource =
            TilgangsrestriksjonResource().apply {
                systemId = Identifikator().apply { identifikatorverdi = "tilgangsId" }
            }

        whenever(administrativEnhetResourceCache.get("adminEnhetKey")).thenReturn(administrativEnhetResource)
        whenever(tilgangsrestriksjonResourceCache.get("tilgangsKey")).thenReturn(tilgangsrestriksjonResource)

        val sakDto =
            SakDto
                .builder()
                .administrativEnhet("adminEnhetKey")
                .skjerming(SkjermingDto.builder().tilgangsrestriksjon("tilgangsKey").build())
                .build()
        val caseSearchParametersDto =
            CaseSearchParametersDto
                .builder()
                .administrativEnhet(true)
                .tilgangsrestriksjon(true)
                .build()

        val result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto)

        assertThat(result).isEqualTo("administrativenhet eq 'adminEnhetId' and tilgangskode eq 'tilgangsId'")
    }

    @Test
    fun `creates a filter query param value for klassering klassifikasjonssystem`() {
        val klassifikasjonssystemResource =
            KlassifikasjonssystemResource().apply {
                systemId =
                    Identifikator().apply { identifikatorverdi = "klassifikasjonssystemId" }
            }
        whenever(
            klassifikasjonssystemResourceCache.get("klassifikasjonssystemKey"),
        ).thenReturn(klassifikasjonssystemResource)

        val sakDto =
            SakDto
                .builder()
                .klasse(
                    listOf(
                        KlasseDto
                            .builder()
                            .rekkefolge(1)
                            .klassifikasjonssystem("klassifikasjonssystemKey")
                            .build(),
                    ),
                ).build()
        val caseSearchParametersDto =
            CaseSearchParametersDto
                .builder()
                .klassering(true)
                .klasseringRekkefolge("1")
                .klasseringKlassifikasjonssystem(true)
                .build()

        val result = caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto)

        assertThat(result).isEqualTo("klassifikasjon/primar/ordning eq 'klassifikasjonssystemId'")
    }

    @Test
    fun `given klassering rekkefolge out of bounds, throws KlasseOrderOutOfBoundsException`() {
        val sakDto =
            SakDto
                .builder()
                .klasse(
                    listOf(
                        KlasseDto
                            .builder()
                            .rekkefolge(0)
                            .klasseId("klasseId")
                            .build(),
                    ),
                ).build()
        val caseSearchParametersDto =
            CaseSearchParametersDto
                .builder()
                .klassering(true)
                .klasseringRekkefolge("0")
                .klasseringKlasseId(true)
                .build()

        assertThatThrownBy { caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto) }
            .isInstanceOf(KlasseOrderOutOfBoundsException::class.java)
            .hasMessage("Rekkefolge=0 is out of bounds. Rekkefolge must be 1, 2 or 3.")
    }

    @Test
    fun `given a missing klassering order, throws SearchKlasseOrderNotFoundInCaseException`() {
        val sakDto =
            SakDto
                .builder()
                .klasse(
                    listOf(
                        KlasseDto
                            .builder()
                            .rekkefolge(1)
                            .klasseId("klasseId")
                            .build(),
                    ),
                ).build()
        val caseSearchParametersDto =
            CaseSearchParametersDto
                .builder()
                .klassering(true)
                .klasseringRekkefolge("2")
                .klasseringKlasseId(true)
                .build()

        assertThatThrownBy { caseSearchParametersService.createFilterQueryParamValue(sakDto, caseSearchParametersDto) }
            .isInstanceOf(SearchKlasseOrderNotFoundInCaseException::class.java)
    }
}
